### PR TITLE
Improve performance when querying a population with get() and ids()

### PR DESCRIPTION
**Pull request by GianlucaFicarelli**
Main improvements:
- use pandas series when the type is category, instead of converting to numpy arrays of objects
- avoid creating full numpy arrays of True or False
- quickly return when False or True is found in masks, if possible

See NSETM-2181 for more details.

### CODE REVIEW

**Comment by GianlucaFicarelli** [comment]
Some benchmarks about reducing masks with OR:
```python
import operator
import numpy as np
from functools import reduce

def _wrap_or(x, y):
    return operator.or_(x, y)

np.random.seed(0)
masks = [np.random.choice([True, False], 80_000_000) for i in range(6)]

for i in range(1, 6):
    print(i)
    _ = %timeit reduce(operator.or_, masks[:i])
    _ = %timeit reduce(np.logical_or, masks[:i])
    _ = %timeit reduce(_wrap_or, masks[:i])
    _ = %timeit np.any(masks[:i], axis=0)
    _ = %timeit np.logical_or.reduce(masks[:i])
```
Result:
```
1
335 ns ± 4.62 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
334 ns ± 0.0464 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
296 ns ± 0.0496 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
67.2 ms ± 57.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
67.1 ms ± 44.6 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
2
30.9 ms ± 52.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
30.9 ms ± 49.2 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
30.9 ms ± 32.3 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
113 ms ± 442 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
112 ms ± 214 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
3
43.4 ms ± 53.8 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
63.1 ms ± 23 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
63.2 ms ± 32 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
157 ms ± 942 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
159 ms ± 326 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
4
56.6 ms ± 18.7 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
96.8 ms ± 41.2 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
96.8 ms ± 55.1 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
204 ms ± 2.24 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
204 ms ± 896 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
5
69.3 ms ± 57.7 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
129 ms ± 38.5 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
127 ms ± 30.8 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
247 ms ± 411 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
247 ms ± 807 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Calling `reduce(operator.or_, masks[:i])` is almost always faster.
Looking at the flame graph, I think that in this case numpy is able to use the elision code https://github.com/numpy/numpy/blob/c6e1fd039da2e6127ba0798f327e9e1b122d7534/numpy/core/src/multiarray/temp_elide.c
from https://github.com/numpy/numpy/blob/c6e1fd039da2e6127ba0798f327e9e1b122d7534/numpy/core/src/multiarray/number.c#L592
while in the other cases it's not.

File: bluepysnap/query.py [hunk 1]
```diff
@@ -19,12 +21,57 @@
...
+    filtered_masks = []
+    for mask in masks:
+        if mask is True:
+            return True
+        if mask is False:
+            continue
+        filtered_masks.append(mask)
+    if not filtered_masks:
+        return False
+    # reduce(operator.or_, masks) is faster than np.logical_or.reduce() or np.any()
```
Comment (mgeplf) [inline bluepysnap/query.py:66]: huh; fun

File: bluepysnap/query.py [hunk 2]
```diff
@@ -19,12 +21,57 @@
...
+    return reduce(operator.or_, (np.asarray(m, dtype=bool) for m in filtered_masks))
+
+
 def _complex_query(prop, query):
-    result = np.full(len(prop), True)
+    result = True
     for key, value in query.items():
         if key == REGEX_KEY:
-            result = np.logical_and(result, prop.str.match(value + "\\Z"))
+            result = _logical_and([result, prop.str.match(value + "\\Z")])
```
Comment (mgeplf) [inline bluepysnap/query.py:74]:
```suggestion
            result = _logical_and([result, prop.str.match(value + r"\Z")])
```

**Review by mgeplf — COMMENTED** [review]
seems very nice/reasonable to me

File: bluepysnap/query.py [hunk 3]
```diff
@@ -19,12 +21,57 @@
...
+    for mask in masks:
+        if mask is True:
+            return True
+        if mask is False:
+            continue
+        filtered_masks.append(mask)
+    if not filtered_masks:
+        return False
+    # reduce(operator.or_, masks) is faster than np.logical_or.reduce() or np.any()
+    return reduce(operator.or_, (np.asarray(m, dtype=bool) for m in filtered_masks))
```
Comment (joni-herttuainen) [inline bluepysnap/query.py:67]: Wow, this is lightning fast, tried a few things but just couldn't beat this. Good job!

**Review by joni-herttuainen — APPROVED** [review]
If it was me, I'd add some sanity checks just for the logical functions to check that:
`_logical_and` == `np.logical_and`, and 
`_logical_or`  == `np.logical_or`
but given how many tests they go through already, this feels almost redundant.

**Comment by GianlucaFicarelli** [comment]
> If it was me, I'd add some sanity checks just for the logical functions to check that:

Sure, I'll add some explicit tests for _logical_or and _logical_and.

**Review by joni-herttuainen — APPROVED** [review]

**Pull request merged**